libbpf-rs/src/perf_buffer.rs

```diff
@@ -218,6 +219,31 @@ impl<'b> PerfBuffer<'b> {
             unsafe { libbpf_sys::perf_buffer__buffer_fd(self.ptr, buf_idx as libbpf_sys::size_t) };
         util::parse_ret_i32(ret)
     }
+
+    /// Get a mutable access to the underlying per-cpu ring buffer data, this API can be used
+    /// to implement a different draining mechanisim for the perf buffer data.
+    ///
+    /// Notice: The API will return a mutable reference to the same memory region in case
+    ///         it's called more than once with the same index, this due to it's nature.
```

Comment:
The reason I went with &self, as this would eventually have the same affect on the object as a call to consume/consume_buffer which both get &self and not &mut self.

Reply:
Okay, but `consume`/`consume_buffer` are sound because they don't return mutable references to internal data, `buffer` is not because it does.

Reply:
I think it will be wrong to make this API &mut self, I do see your point in marking it as an unsafe function.

I'll give you an example why I think this will make this API much harder to use, and in case we go down this road
it means we should probably implement all the other helpers Vec has like split_at_mut, I think it's better to think of this API like get_unchecked_mut, perhaps it should be renamed to buffer_unchecked_mut and marked as unsafe.

As for an example, let's pretend we have a structure called RingBuffer, which implements an iterator over the event in the perf buffer, by making this &self mut, it means you can only have a single instance of this RingBuffer, but you have N cores, so you want to create a RingBuffer structure per perf-buffer.

Will you be willing to accept renaming this API to buffer_unchecked_mut and marked as unsafe?

Reply:
> As for an example, let's pretend we have a structure called RingBuffer, which implements an iterator over the event in the perf buffer, by making this &self mut, it means you can only have a single instance of this RingBuffer, but you have N cores, so you want to create a RingBuffer structure per perf-buffer.

Thanks for the example. I certainly can see why you would want this API, I am just trying to figure out if there is a better design possible. In my opinion and assuming I am understanding correctly, the more appropriate way to go about expressing the contract in code would be to have something like:

```rust
impl<'b> PerfBuffer<'b> {
  pub fn into_cpu_buffers(self) -> Vec<CpuBuffer<'b>>;
}
```
The problem will be that there is a bunch of infrastructure that you will need on `CpuBuffer` that currently only exists on `PerfBuffer` and would need to be duplicated in one form or another. In short, it's not trivial to retro-fit that design on the existing one and given that it's more or less a niche use (most users probably won't use that lower level infrastructure), it may not be worth going down that road.

> I think it's better to think of this API like get_unchecked_mut, perhaps it should be renamed to buffer_unchecked_mut and marked as unsafe.

Do keep in mind that `get_unchecked_mut` still has a `&mut self` receiver, which eliminates a slew of intricacies in its usage.

> Will you be willing to accept renaming this API to buffer_unchecked_mut and marked as unsafe?

That would definitely be better. We will need to slap a bunch of warnings on it. And again, a test case is still missing.

Reply:
> Do keep in mind that get_unchecked_mut still has a &mut self receiver, which eliminates a slew of intricacies in its usage.

Ye, you are right my bad, that why i edited my original reply :P, but it sounds like you understand my point, as for your suggestion yes, that's possible but like you said it will still be a bit annoying as there are stuff that in PerfBuffer which would be useful to keep.

As for the name of the API if you will review my (edited reply) I suggest it would be buffer_mut since the word unchecked is usually meant for index checking, where here we still do them.
so it would look like
`pub unsafe fn buffer_mut(&self, buf_idx: usize) -> Result<&mut [u8]>`

What kind of test did you have in mind? and any chance we have some test I can just extend or use as a sample?

Reply:
> As for the name of the API if you will review my (edited reply) I suggest it would be buffer_mut since the word unchecked is usually meant for index checking, where here we still do them.
> so it would look like
> `pub unsafe fn buffer_mut(&self, buf_idx: usize) -> Result<&mut [u8]>`

Yep, I think that sounds good.

> What kind of test did you have in mind? and any chance we have some test I can just extend or use as a sample?

Anything that demonstrates that `buffer_mut` works as expected will do. So I'd think that writing a value to a ring buffer from a BPF program and then scanning all buffers from user space to make sure it arrived somewhere, perhaps? Should be possible to distill parts of what `capable` is doing, for example, and enhance it a bit before putting it into `libbpf-rs/tests/test.rs`.

Reply:
I have submitted a test patch, where I produce a know value and look for it in the perf buffer